### PR TITLE
Preprocessing changes

### PR DESCRIPTION
**Pull request by evcano**
Hi @bch0w,

As I mentioned in issue #152, I made the following changes to the preprocessing module:

1. **Synthetics and observations can have a different format**. This change was on the default preprocessing module. However, I changed the name of the variable <data_format> in the solver and pyaflowa modules in order to keep consistency. I also updated the tests, examples, and documentation to account for my changes.
2. **Observations can have SAC format**
3. **Added <unit_output> variable to default preprocessing module**
4. **Added hard check: the observed and synthetic data format must be correct**
5. **Added hard check: synthetic and observations files must match**
6. **Added function to check that all required adjoint sources exist**
7. **Fixed: Adjoint traces were incorrectly written**. Adjoint traces were initialized as a copy of the obspy Stream containing the synthetic trace. Then, the adjoint trace was written in a new property of the Stream called "data" instead of being written on the data property of the Trace object contained in the Stream. This resulted in the synthetic trace being written instead of the adjoint trace.

Feel free to modify what you consider necessary.

Best,
Eduardo.

### CODE REVIEW

**Comment by bch0w** [comment]
Hi @evcano, thanks for the PR. Just FYI I'm out of town the next few days but will check this out later this week!

**Comment by evcano** [comment]
Hi @bch0w, I think I fixed an important bug in the default preprocessing module. Please see point 7 of my PR.

**Comment by bch0w** [comment]
Hi @evcano, wow! Great catch. That's a pretty sneaky bug, I'm surprised it worked and provided usable results in the examples. Thanks very much for going through the preprocess module with such detail, it's incredibly useful!

Sorry it's taken me so long to get to this PR, I got caught up in some other work. I've had a look through all the proposed changes and everything looks good to me. Just want to run a few example problems on my local to test it out and then I will merge the PR.

**Comment by evcano** [comment]
Hi @bch0w, no problem! Take your time

File: seisflows/preprocess/default.py
Comment (bch0w) [inline seisflows/preprocess/default.py:31]: `syn_data_format` is shared by preprocess.default and solver.specfem and both classes list `syn_data_format` in their class docstring, which causes it to show up twice in the dynamically generated parameter file created during `seisflows configure`.

```yaml
#
#    Solver SPECFEM
#    --------------
#    Defines foundational structure for Specfem-based solver module.
#    Generalized SPECFEM interface to manipulate SPECFEM2D/3D/3D_GLOBE w/ Python
#
#    Parameters
#    ----------
#    :type syn_data_format: str
#    :param syn_data_format: data format for reading synthetic traces into memory.
#        Available: ['SU': seismic unix format, 'ASCII': human-readable ascii]
...
#    Default Preprocess
#    ------------------
#    Data processing for seismic traces, with options for data misfit,
#    filtering, normalization and muting.
#
#    Parameters
#    ----------
#    :type syn_data_format: str
#    :param syn_data_format: data format for reading synthetic traces into
#        memory. Shared with solver module. Available formats: 'su', 'ascii'
```


Could you move one of the docstrings for `syn_data_format` into the __init__() function of its respective class. This allows it to still be available in the source code, but to not show up twice in the parameter file.

I would suggest moving this one, since `syn_data_format` is more relevant to the solver, but I leave it up to you!

**Review by bch0w — CHANGES_REQUESTED** [review]
Everything looks great @evcano! Just one suggestion about amending a docstring to avoid repeated parameters in the dynamically generated parameter file. Once that is done, I am happy to merge this into the devel branch. Thanks!

**Comment by bch0w** [comment]
Hi @evcano, I pushed the suggestion myself since it was small. Hope that's alright. Thanks again for the contribution!

**Comment by evcano** [comment]
Of course, thanks for modifying it for me. I am glad to help!

**Pull request merged**
